### PR TITLE
[SPARK-33860][SQL] Make CatalystTypeConverters.convertToCatalyst match special Array value

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Add some case to match Array whose element type is primitive.

### Why are the changes needed?

We will get exception when use `Literal.create(Array(1, 2, 3), ArrayType(IntegerType))` .
```
Exception in thread "main" java.lang.IllegalArgumentException: requirement failed: Literal must have a corresponding value to array<int>, but class int[] found.
	at scala.Predef$.require(Predef.scala:281)
	at org.apache.spark.sql.catalyst.expressions.Literal$.validateLiteralValue(literals.scala:215)
	at org.apache.spark.sql.catalyst.expressions.Literal.<init>(literals.scala:292)
	at org.apache.spark.sql.catalyst.expressions.Literal$.create(literals.scala:140)
```
And same problem with other array whose element is primitive.

### Does this PR introduce _any_ user-facing change?

Yes.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @maropu @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:462]: How does this PR fixes `Literal(Array(1, 2, 3)) == Literal.create(Array(1, 2, 3), ArrayType(IntegerType))` case?

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:462]: `Array[Any]` doesn't match the `Array(1, 2, 3)`. This PR use the `Array[_]` to match it.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala:338]: Is this a user-facing bug? If so, its better to add end-2-end tests, too.

Comment (ulysses-you) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/LiteralExpressionSuite.scala:338]: I think it's not a user-facing bug. User usually use the `Literal.apply` code path and the `Literal.create` is used by Spark self.

**Review by maropu — APPROVED** [review]
Looks a minor internal bug and fixing it looks fine to me. I'll leave this to @HyukjinKwon or the other committers.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala
Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:462]: It already works without this change.

```scala
scala> Literal(Array(1, 2, 3))
res1: org.apache.spark.sql.catalyst.expressions.Literal = [1,2,3]
```

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:462]: Seems missed a param `ArrayType(IntegerType)`, they use the different code path.

Comment (HyukjinKwon) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/CatalystTypeConverters.scala:462]: Oh, gotya.

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.1, branch-3.0 and branch-2.4.

**Comment by HyukjinKwon** [comment]
It has a conflict with branch-2.4 but I think we don't have to bother.

**Comment by ulysses-you** [comment]
thanks for merging !

**Comment by dongjoon-hyun** [comment]
Hi, @ulysses-you and @HyukjinKwon . This seems to break `branch-3.0`. Could you take a look at that?
- https://github.com/apache/spark/runs/1593517087
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test%20(Dashboard)/job/spark-branch-3.0-test-sbt-hadoop-3.2-hive-2.3/1056/testReport/org.apache.spark.sql.catalyst.expressions/LiteralExpressionSuite/convertToCatalyst_match_special_Array_value/

**Comment by HyukjinKwon** [comment]
Thanks @dongjoon-hyun. I will revert in branch-3.0 for now.

**Comment by ulysses-you** [comment]
@dongjoon-hyun @HyukjinKwon I think it's because of  [#30295](https://github.com/apache/spark/pull/30295) not backport to branch-3.0. If we need backport this PR then we need backport [#30295](https://github.com/apache/spark/pull/30295)  too.

**Comment by HyukjinKwon** [comment]
I think we can just leave it, it's fine.

**Comment by ulysses-you** [comment]
+1 for this.

**Pull request closed without merging**
